src/runtime/runtime_state.cs

```diff
@@ -0,0 +1,236 @@
+            unsafe
+            {
+                var head = (PyGC_Head*)dummyGCHead;
+                head->gc.gc_next = dummyGCHead;
+                head->gc.gc_prev = dummyGCHead;
+                head->gc.gc_refs = IntPtr.Zero;
+            }
+            {
+                var pyDummyGC = PyLong_FromVoidPtr(dummyGCHead);
+                int res = PySys_SetObject("dummy_gc", pyDummyGC);
+                PythonException.ThrowIfIsNotZero(res);
+                XDecref(pyDummyGC);
+
+                try
+                {
+                    res = PySys_SetObject("initial_modules", modules);
+                    PythonException.ThrowIfIsNotZero(res);
+                }
+                finally
+                {
+                    XDecref(modules);
+                }
+
+                if (ShouldRestoreObjects)
+                {
+                    AddObjPtrToSet(objs, modules);
+                    try
+                    {
+                        res = PySys_SetObject("initial_objs", objs);
+                        PythonException.ThrowIfIsNotZero(res);
+                    }
+                    finally
+                    {
+                        XDecref(objs);
+                    }
+                }
+            }
+        }
+
+        public static void Restore()
+        {
+            var dummyGCAddr = PySys_GetObject("dummy_gc");
+            if (dummyGCAddr == IntPtr.Zero)
+            {
+                throw new Exception("Runtime state have not set");
+            }
+            var dummyGC = PyLong_AsVoidPtr(dummyGCAddr);
+            ResotreModules(dummyGC);
+            if (ShouldRestoreObjects)
+            {
+                RestoreObjects(dummyGC);
+            }
+        }
+
+        private static void ResotreModules(IntPtr dummyGC)
+        {
+            var intialModules = PySys_GetObject("initial_modules");
+            Debug.Assert(intialModules != null);
+            var modules = PyImport_GetModuleDict();
+            foreach (var name in GetModuleNames())
+            {
+                if (PySet_Contains(intialModules, name) == 1)
+                {
+                    continue;
+                }
+                var module = PyDict_GetItem(modules, name);
+
+                if (UseDummyGC && _PyObject_GC_IS_TRACKED(module))
+                {
+                    ExchangeGCChain(module, dummyGC);
+                }
+                if (PyDict_DelItem(modules, name) != 0)
+                {
+                    PyErr_Print();
+                }
+            }
+        }
+
+        private static void RestoreObjects(IntPtr dummyGC)
+        {
+            if (!UseDummyGC)
+            {
+                throw new Exception("To prevent crash by _PyObject_GC_UNTRACK in Python internal, UseDummyGC should be enabled when using ResotreObjects");
+            }
+            var intialObjs = PySys_GetObject("initial_objs");
+            Debug.Assert(intialObjs != null);
```

Comment:
Please, use explicit type here instead of `var`.
I am unsure, but I'd guess it is `IntPtr`, in which case `initialObjs != null` is always `true`, because `IntPtr` is a value type. In this case you need to compare to `IntPtr.Zero`.